File: Python/OneLang/One/Transforms/InferTypesPlugins/Helpers/GenericsResolver.py
from OneLangStdLib import *
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.AstTypes as astTypes
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.Interfaces as ints

class GenericsResolver:
    def __init__(self):
        self.resolution_map = Map()
    
    @classmethod
    def from_object(cls, object):
        resolver = GenericsResolver()
        resolver.collect_class_generics_from_object(object)
        return resolver
    
    def add_resolution(self, type_var_name, actual_type):
        prev_res = self.resolution_map.get(type_var_name)
        if prev_res != None and not astTypes.TypeHelper.equals(prev_res, actual_type):
            raise Error(f'''Resolving \'{type_var_name}\' is ambiguous, {prev_res.repr()} <> {actual_type.repr()}''')
        self.resolution_map.set(type_var_name, actual_type)
    
    def collect_from_method_call(self, method_call):
        if len(method_call.type_args) == 0:
            return
        if len(method_call.type_args) != len(method_call.method.type_arguments):
            raise Error(f'''Expected {len(method_call.method.type_arguments)} type argument(s) for method call, but got {len(method_call.type_args)}''')
        i = 0
        
        while i < len(method_call.type_args):
            self.add_resolution(method_call.method.type_arguments[i], method_call.type_args[i])
            i = i + 1
    
    def collect_class_generics_from_object(self, actual_object):
        actual_type = actual_object.get_type()
        if isinstance(actual_type, astTypes.ClassType):
            if not self.collect_resolutions_from_actual_type(actual_type.decl.type, actual_type):
                pass
        elif isinstance(actual_type, astTypes.InterfaceType):
            if not self.collect_resolutions_from_actual_type(actual_type.decl.type, actual_type):
                pass
        else:
            raise Error(f'''Expected ClassType or InterfaceType, got {(actual_type.repr() if actual_type != None else "<null>")}''')
    
    def collect_resolutions_from_actual_type(self, generic_type, actual_type):
        if not astTypes.TypeHelper.is_generic(generic_type):
            return True
        if isinstance(generic_type, astTypes.GenericsType):
            self.add_resolution(generic_type.type_var_name, actual_type)
            return True
        elif isinstance(generic_type, astTypes.ClassType) and isinstance(actual_type, astTypes.ClassType) and generic_type.decl == actual_type.decl:
            if len(generic_type.type_arguments) != len(actual_type.type_arguments):
                raise Error(f'''Same class ({generic_type.repr()}) used with different number of type arguments ({len(generic_type.type_arguments)} <> {len(actual_type.type_arguments)})''')
            return ArrayHelper.every(lambda x, i: self.collect_resolutions_from_actual_type(x, actual_type.type_arguments[i]), generic_type.type_arguments)
        elif isinstance(generic_type, astTypes.InterfaceType) and isinstance(actual_type, astTypes.InterfaceType) and generic_type.decl == actual_type.decl:
            if len(generic_type.type_arguments) != len(actual_type.type_arguments):
                raise Error(f'''Same class ({generic_type.repr()}) used with different number of type arguments ({len(generic_type.type_arguments)} <> {len(actual_type.type_arguments)})''')
            return ArrayHelper.every(lambda x, i: self.collect_resolutions_from_actual_type(x, actual_type.type_arguments[i]), generic_type.type_arguments)
        elif isinstance(generic_type, astTypes.LambdaType) and isinstance(actual_type, astTypes.LambdaType):
            if len(generic_type.parameters) != len(actual_type.parameters):
                raise Error(f'''Generic lambda type has {len(generic_type.parameters)} parameters while the actual type has {len(actual_type.parameters)}''')
            params_ok = ArrayHelper.every(lambda x, i: self.collect_resolutions_from_actual_type(x.type, actual_type.parameters[i].type), generic_type.parameters)
            result_ok = self.collect_resolutions_from_actual_type(generic_type.return_type, actual_type.return_type)
            return params_ok and result_ok
        elif isinstance(generic_type, astTypes.EnumType) and isinstance(actual_type, astTypes.EnumType) and generic_type.decl == actual_type.decl:
            pass
        elif isinstance(generic_type, astTypes.AnyType) or isinstance(actual_type, astTypes.AnyType):
            pass
        else:
            raise Error(f'''Generic type {generic_type.repr()} is not compatible with actual type {actual_type.repr()}''')
        return False
    
    def resolve_type(self, type, must_resolve_all_generics):
        if isinstance(type, astTypes.GenericsType):
            resolved_type = self.resolution_map.get(type.type_var_name)
            if resolved_type == None and must_resolve_all_generics:
                raise Error(f'''Could not resolve generics type: {type.repr()}''')
            return resolved_type if resolved_type != None else type
        elif isinstance(type, astTypes.ClassType):
            return astTypes.ClassType(type.decl, list(map(lambda x: self.resolve_type(x, must_resolve_all_generics), type.type_arguments)))
        elif isinstance(type, astTypes.InterfaceType):
            return astTypes.InterfaceType(type.decl, list(map(lambda x: self.resolve_type(x, must_resolve_all_generics), type.type_arguments)))
        elif isinstance(type, astTypes.LambdaType):
            return astTypes.LambdaType(list(map(lambda x: types.MethodParameter(x.name, self.resolve_type(x.type, must_resolve_all_generics), x.initializer, None), type.parameters)), self.resolve_type(type.return_type, must_resolve_all_generics))
        else:
            return type

File: PHP/src/OneLang/One/Transforms/InferTypesPlugins/Helpers/GenericsResolver.php
<?php

namespace OneLang\One\Transforms\InferTypesPlugins\Helpers\GenericsResolver;

use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\IMethodCallExpression;
use OneLang\One\Ast\AstTypes\ClassType;
use OneLang\One\Ast\AstTypes\GenericsType;
use OneLang\One\Ast\AstTypes\InterfaceType;
use OneLang\One\Ast\AstTypes\LambdaType;
use OneLang\One\Ast\AstTypes\EnumType;
use OneLang\One\Ast\AstTypes\AnyType;
use OneLang\One\Ast\AstTypes\TypeHelper;
use OneLang\One\Ast\Types\MethodParameter;
use OneLang\One\Ast\Interfaces\IType;

class GenericsResolver {
    public $resolutionMap;
    
    function __construct()
    {
        $this->resolutionMap = new \OneLang\Core\Map();
    }
    
    static function fromObject($object) {
        $resolver = new GenericsResolver();
        $resolver->collectClassGenericsFromObject($object);
        return $resolver;
    }
    
    function addResolution($typeVarName, $actualType) {
        $prevRes = $this->resolutionMap->get($typeVarName);
        if ($prevRes !== null && !TypeHelper::equals($prevRes, $actualType))
            throw new \OneLang\Core\Error("Resolving '" . $typeVarName . "' is ambiguous, " . $prevRes->repr() . " <> " . $actualType->repr());
        $this->resolutionMap->set($typeVarName, $actualType);
    }
    
    function collectFromMethodCall($methodCall) {
        if (count($methodCall->typeArgs) === 0)
            return;
        if (count($methodCall->typeArgs) !== count($methodCall->method->typeArguments))
            throw new \OneLang\Core\Error("Expected " . count($methodCall->method->typeArguments) . " type argument(s) for method call, but got " . count($methodCall->typeArgs));
        for ($i = 0; $i < count($methodCall->typeArgs); $i++)
            $this->addResolution($methodCall->method->typeArguments[$i], $methodCall->typeArgs[$i]);
    }
    
    function collectClassGenericsFromObject($actualObject) {
        $actualType = $actualObject->getType();
        if ($actualType instanceof ClassType) {
            if (!$this->collectResolutionsFromActualType($actualType->decl->type, $actualType)) { }
        }
        else if ($actualType instanceof InterfaceType) {
            if (!$this->collectResolutionsFromActualType($actualType->decl->type, $actualType)) { }
        }
        else
            throw new \OneLang\Core\Error("Expected ClassType or InterfaceType, got " . ($actualType !== null ? $actualType->repr() : "<null>"));
    }
    
    function collectResolutionsFromActualType($genericType, $actualType) {
        if (!TypeHelper::isGeneric($genericType))
            return true;
        if ($genericType instanceof GenericsType) {
            $this->addResolution($genericType->typeVarName, $actualType);
            return true;
        }
        else if ($genericType instanceof ClassType && $actualType instanceof ClassType && $genericType->decl === $actualType->decl) {
            if (count($genericType->typeArguments) !== count($actualType->typeArguments))
                throw new \OneLang\Core\Error("Same class (" . $genericType->repr() . ") used with different number of type arguments (" . count($genericType->typeArguments) . " <> " . count($actualType->typeArguments) . ")");
            return \OneLang\Core\ArrayHelper::every($genericType->typeArguments, function ($x, $i) use ($actualType) { return $this->collectResolutionsFromActualType($x, $actualType->typeArguments[$i]); });
        }
        else if ($genericType instanceof InterfaceType && $actualType instanceof InterfaceType && $genericType->decl === $actualType->decl) {
            if (count($genericType->typeArguments) !== count($actualType->typeArguments))
                throw new \OneLang\Core\Error("Same class (" . $genericType->repr() . ") used with different number of type arguments (" . count($genericType->typeArguments) . " <> " . count($actualType->typeArguments) . ")");
            return \OneLang\Core\ArrayHelper::every($genericType->typeArguments, function ($x, $i) use ($actualType) { return $this->collectResolutionsFromActualType($x, $actualType->typeArguments[$i]); });
        }
        else if ($genericType instanceof LambdaType && $actualType instanceof LambdaType) {
            if (count($genericType->parameters) !== count($actualType->parameters))
                throw new \OneLang\Core\Error("Generic lambda type has " . count($genericType->parameters) . " parameters while the actual type has " . count($actualType->parameters));
            $paramsOk = \OneLang\Core\ArrayHelper::every($genericType->parameters, function ($x, $i) use ($actualType) { return $this->collectResolutionsFromActualType($x->type, $actualType->parameters[$i]->type); });
            $resultOk = $this->collectResolutionsFromActualType($genericType->returnType, $actualType->returnType);
            return $paramsOk && $resultOk;
        }
        else if ($genericType instanceof EnumType && $actualType instanceof EnumType && $genericType->decl === $actualType->decl) { }
        else if ($genericType instanceof AnyType || $actualType instanceof AnyType) { }
        else
            throw new \OneLang\Core\Error("Generic type " . $genericType->repr() . " is not compatible with actual type " . $actualType->repr());
        return false;
    }
    
    function resolveType($type, $mustResolveAllGenerics) {
        if ($type instanceof GenericsType) {
            $resolvedType = $this->resolutionMap->get($type->typeVarName);
            if ($resolvedType === null && $mustResolveAllGenerics)
                throw new \OneLang\Core\Error("Could not resolve generics type: " . $type->repr());
            return $resolvedType !== null ? $resolvedType : $type;
        }
        else if ($type instanceof ClassType)
            return new ClassType($type->decl, array_map(function ($x) use ($mustResolveAllGenerics) { return $this->resolveType($x, $mustResolveAllGenerics); }, $type->typeArguments));
        else if ($type instanceof InterfaceType)
            return new InterfaceType($type->decl, array_map(function ($x) use ($mustResolveAllGenerics) { return $this->resolveType($x, $mustResolveAllGenerics); }, $type->typeArguments));
        else if ($type instanceof LambdaType)
            return new LambdaType(array_map(function ($x) use ($mustResolveAllGenerics) { return new MethodParameter($x->name, $this->resolveType($x->type, $mustResolveAllGenerics), $x->initializer, null); }, $type->parameters), $this->resolveType($type->returnType, $mustResolveAllGenerics));
        else
            return $type;
    }
}
